Adds E2E coverage reporting

m1kola opened the pull request:
# Description

Adds E2E coverage reporting.

## Reviewer Checklist

- [ ] API Go Documentation
- [ ] Tests: Unit Tests (and E2E Tests, if appropriate)
- [ ] Comprehensive Commit Messages
- [ ] Links to related GitHub Issue(s)

perdasilva commented on Makefile:
```diff
@@ -103,7 +105,16 @@ test-unit: $(SETUP_ENVTEST) ## Run the unit tests
+e2e: KUSTOMIZE_BUILD_DIR=config/e2e
+e2e: GO_BUILD_FLAGS=-cover
+e2e: run kind-load-test-artifacts test-e2e ## Run e2e test suite on local kind cluster
+	# Coverage-instrumented binary produces coverage on termination, so we scale down the manager before gathering the coverage
```
great use of comments ^^ thank you!

m1kola commented:
@perdasilva thanks for the review.

> What do you think is missing before we can take it off draft state?

There were few things which I wanted to change before marking ready for review:

* I didn't want to have a bunch of `kubectl` commands in the `Makefile` They are now in `./hack/e2e-coverage.sh`.
* I didn't like the very long json in `--overrides` in `kubectl run`. I moved pod creation into `config/e2e/manager_e2e_coverage_copy_pod.yaml`.
* I had to change github action to make sure that we collect coverage even if tests fail. In the previous version coverage was only collected on success due to `make` aborting execution on first non-zero exit code. I added `-k`/`--keep-going` flag to `make -k e2e` which will continue make build, but will return non-zero exit code in case of failures.

The PR should be ready now. I'm just checking few things before I actually take it off the draft state.

ncdc commented on .github/workflows/e2e.yaml:
```diff
@@ -22,4 +22,11 @@ jobs:
     - name: Run e2e tests
       run: |
-        make e2e
+        make -k e2e
```
Mind commenting why `-k` is here? I liked your explanation in the comments and think it would be worth putting in here.

ncdc commented on Makefile:
```diff
@@ -103,7 +105,13 @@ test-unit: $(SETUP_ENVTEST) ## Run the unit tests
 	eval $$($(SETUP_ENVTEST) use -p env $(ENVTEST_VERSION)) && go test -tags $(GO_BUILD_TAGS) -count=1 -short $(UNIT_TEST_DIRS) -coverprofile cover.out
 
 e2e: KIND_CLUSTER_NAME=operator-controller-e2e
-e2e: run kind-load-test-artifacts test-e2e kind-cluster-cleanup ## Run e2e test suite on local kind cluster
```
I see that the help text `## Run e2e test suite on local kind cluster` is gone - did you mean to remove it?

m1kola replied:
Good catch! Didn't mean to remove help comment.

ncdc commented on hack/e2e-coverage.sh:
```diff
@@ -0,0 +1,33 @@
+kubectl -n "$OPERATOR_CONTROLLER_NAMESPACE" scale deployment/"$OPERATOR_CONTROLLER_MANAGER_DEPLOYMENT_NAME" --replicas=0
+
+# Wait for the copy pod to be ready
+kubectl -n "$OPERATOR_CONTROLLER_NAMESPACE" wait --for=condition=ready pod "$COPY_POD_NAME"
```
Is it possible for the copy pod to get scheduled and be ready before the manager pod, i.e. "locking" the PVC and preventing the manager pod from starting?

ncdc replied:
Thanks, I was incorrect in my thinking that ReadWriteOnce meant it would only bind to 1 pod at a time.
